server/coordinator/node_picker.go

```diff
@@ -4,53 +4,91 @@ package coordinator
+	PickNode(ctx context.Context, shardIDs []storage.ShardID, shardTotalNum uint32, registerNodes []metadata.RegisteredNode) (map[storage.ShardID]metadata.RegisteredNode, error)
+}
+type UniformityConsistentHashNodePicker struct {
+	logger *zap.Logger
 }
 
-type ConsistentHashNodePicker struct {
-	logger       *zap.Logger
-	hashReplicas int
+func NewUniformityConsistentHashNodePicker(logger *zap.Logger) NodePicker {
+	return &UniformityConsistentHashNodePicker{logger: logger}
 }
 
-func NewConsistentHashNodePicker(logger *zap.Logger, hashReplicas int) NodePicker {
-	return &ConsistentHashNodePicker{hashReplicas: hashReplicas, logger: logger}
+type nodeMember string
+
+func (m nodeMember) String() string {
+	return string(m)
 }
 
-func (p *ConsistentHashNodePicker) PickNode(_ context.Context, shardID storage.ShardID, registerNodes []metadata.RegisteredNode) (metadata.RegisteredNode, error) {
+type hasher struct{}
+
+func (h hasher) Sum64(data []byte) uint64 {
+	return murmur3.Sum64(data)
+}
+
+func (p *UniformityConsistentHashNodePicker) PickNode(_ context.Context, shardIDs []storage.ShardID, shardTotalNum uint32, registerNodes []metadata.RegisteredNode) (map[storage.ShardID]metadata.RegisteredNode, error) {
 	now := time.Now()
 
-	hashRing := hash.New(p.hashReplicas, nil)
-	aliveNodeNumber := 0
+	result := make(map[storage.ShardID]metadata.RegisteredNode, len(registerNodes))
+	aliveNodes := make([]metadata.RegisteredNode, 0, len(registerNodes))
 	for _, registerNode := range registerNodes {
 		if !registerNode.IsExpired(now) {
-			hashRing.Add(registerNode.Node.Name)
-			aliveNodeNumber++
+			aliveNodes = append(aliveNodes, registerNode)
 		}
 	}
+	if len(aliveNodes) == 0 {
+		return map[storage.ShardID]metadata.RegisteredNode{}, errors.WithMessage(ErrPickNode, "no alive node in cluster")
+	}
 
-	if hashRing.IsEmpty() {
-		return metadata.RegisteredNode{}, errors.WithMessage(ErrNodeNumberNotEnough, "at least one online nodes is required")
+	mems := make([]hash.Member, 0, len(aliveNodes))
+	for _, node := range aliveNodes {
+		mems = append(mems, nodeMember(node.Node.Name))
 	}
 
-	pickedNodeName := hashRing.Get(strconv.Itoa(int(shardID)))
-	p.logger.Debug("ConsistentHashNodePicker pick result", zap.Uint64("shardID", uint64(shardID)), zap.String("node", pickedNodeName), zap.Int("nodeNumber", aliveNodeNumber))
+	consistentConfig := hash.Config{
+		PartitionCount:    int(shardTotalNum),
+		ReplicationFactor: int(shardTotalNum),
+		Load:              1,
+		Hasher:            hasher{},
+	}
+	c := hash.NewUniformityHash(mems, consistentConfig)
+	for partID := 0; partID < consistentConfig.PartitionCount; partID++ {
+		mem := c.GetPartitionOwner(partID)
+		nodeName := mem.String()
 
-	for i := 0; i < len(registerNodes); i++ {
-		if registerNodes[i].Node.Name == pickedNodeName {
-			return registerNodes[i], nil
+		if contains(storage.ShardID(partID), shardIDs) {
```

Comment:
When partID will not be in shardIDs?

Reply:
It is impossible in theory, but here to be on the safe side, a check is done.